src/devices/cpu/m6809/6x09dasm.h

```diff
@@ -67,33 +68,46 @@ class m6x09_base_disassembler : public util::disasm_interface
 		IMM_RR,             // Register-to-register
 		IMM_BW,             // Bitwise operations (6309 only)
 		IMM_TFM,            // Transfer from memory (6309 only)
-		PG1,                // Switch to page 1 opcodes
-		PG2                 // Switch to page 2 opcodes
+		PG2,                // Switch to page 2 opcodes
+		PG3                 // Switch to page 3 opcodes
 	};
 
 	// Opcode structure
 	class opcodeinfo
 	{
 	public:
-		constexpr opcodeinfo(uint16_t opcode, uint8_t length, const char *name, m6x09_addressing_mode mode, m6x09_instruction_level level, unsigned flags = 0)
-			: m_opcode(opcode), m_length(length), m_mode(mode), m_level(level), m_flags(flags), m_name(name)
+		constexpr opcodeinfo(uint16_t opcode, uint8_t operand_length, const char *name, m6x09_addressing_mode mode, uint32_t level, unsigned flags = 0)
+			: m_opcode(opcode), m_operand_length(operand_length), m_mode(mode), m_level(level), m_flags(flags), m_name(name)
 		{
 		}
 
 		uint16_t opcode() const { return m_opcode; }
-		uint8_t length() const { return m_length; }
+		uint8_t operand_length() const { return m_operand_length; }
 		m6x09_addressing_mode mode() const { return m_mode; }
-		m6x09_instruction_level level() const { return m_level; }
+		uint32_t level() const { return m_level; }
 		unsigned flags() const { return m_flags; }
 		const char *name() const { return m_name; }
 
+		bool operator < (const opcodeinfo & rhs) const
+		{
+			return (opcode() < rhs.opcode());
+		}
+
+		struct compare
+		{
+			using is_transparent = void;
+			bool operator()(opcodeinfo const& lhs, opcodeinfo const& rhs) const;
+			bool operator()(uint16_t opcode, opcodeinfo const& rhs) const;
+			bool operator()(opcodeinfo const& lhs, uint16_t opcode) const;
+		};
+
 	private:
-		uint16_t                m_opcode;       // 8-bit opcode value
-		uint8_t                 m_length;       // Opcode length in bytes
-		m6x09_addressing_mode   m_mode : 6;     // Addressing mode
-		m6x09_instruction_level m_level : 2;    // General, or 6309 only?
-		unsigned                m_flags;        // Disassembly flags
-		const char *            m_name;         // Opcode name
+		uint16_t                m_opcode;         // 8-bit opcode value
+		uint8_t                 m_operand_length; // Opcode length in bytes
+		m6x09_addressing_mode   m_mode : 6;       // Addressing mode
+		uint32_t                m_level;          // General, or 6309 only?
+		unsigned                m_flags;          // Disassembly flags
+		const char *            m_name;           // Opcode name
```

Comment:
The bitfields were being used in an attempt to encourage the compiler to pack the two fields into a single byte so they’d get packed into a word with the opcode and length.  That won’t happen any more, so there’s no point giving a field width.

You can specify a storage class for `m6x09_addressing_mode` at declaration if you want to ensure it’s stored as a byte.

Also, is there any point having `m_level` take a whole 32 bits?  Do you expect to have that many instruction categories?

Reply:
Addressed. Committed.